corgi/tasks/prod_defs.py

```diff
@@ -67,6 +69,89 @@ def update_products() -> None:
                 parse_product_version(pd_product_version, product, product_node)
 
 
+def _find_by_cpe(cpe_patterns: list[str]) -> list[str]:
+    """Given a list of CPE patterns find all the CPEs from ET Variants which match it"""
+    if not cpe_patterns:
+        return []
+    regex_query = Q()
+    for pattern in cpe_patterns:
+        regex = pattern.replace(".", "\\.").replace("*", ".*")
+        regex_query |= Q(cpe__regex=regex)
+    return list(
+        CollectorErrataProductVariant.objects.filter(live=True)
+        .filter(regex_query)
+        .values_list("cpe", flat=True)
+    )
+
+
+def _match_stream_version_to_cpe_version(
+    version_streams: QuerySet[ProductStream],
+    cpe_version: str,
+    cpe: str,
+    el_by_cpe: defaultdict[str, list],
+    cpes_by_stream: defaultdict[str, list],
+):
+    """Iterate product stream versions ordered by length descending, looking for exact matches to
+    the stream's parent ProductVersion cpe. If a match is found, store the match along with any
+    el suffixes."""
+    for version_stream in version_streams:
+        if version_stream.version.removesuffix(".z") == cpe_version:
+            if cpe in el_by_cpe:
+                for el in el_by_cpe[cpe]:
+                    cpes_by_stream[version_stream.name].append(cpe + "::" + el)
+            else:
+                cpes_by_stream[version_stream.name].append(cpe)
+
+
+def _split_cpe_version(cpe_part: str, versions_by_cpe: defaultdict[str, set]):
+    version_split = cpe_part.rsplit(":", 1)
+    cpe_name = version_split[0]
+    cpe_version = version_split[-1]
+    versions_by_cpe[cpe_name].add(cpe_version)
+
+
+def _split_cpe(cpe: str, versions_by_cpe: defaultdict[str, set], el_by_cpe: defaultdict[str, list]):
+    """Splits a cpe such as 'cpe:/a:redhat:openshift_ironic:4.13:el9' into
+    'cpe:/a:redhat:openshift_ironic', '4.13', and 'el9'
+    It builds 2 dictionaries, one with a set of versions grouped by cpe_name and another
+    with el suffixes grouped by 'cpe_name + version'.
+    We do this so that we can iterate versions looking for matches without el suffixes. Once
+    we find a match, we add back the suffixes"""
+    variant_split = cpe.rsplit("::")
+    if len(variant_split) == 2:
+        el_part = variant_split[1]
+        cpe_part = variant_split[0]
+        el_by_cpe[cpe_part].append(el_part)
+        _split_cpe_version(cpe_part, versions_by_cpe)
+    elif len(variant_split) == 1:
+        _split_cpe_version(cpe, versions_by_cpe)
+    else:
+        raise ValueError(f"More than one el qualifier '::' in {cpe}")
+
+
+def match_stream_cpes(product_version: ProductVersion) -> None:
+    """Given a Product Version with Product Stream foreign keys and cpes_matching_patterns
+    This will match those cpes to the stream children using the stream versions"""
+    versions_by_cpe: defaultdict[str, set] = defaultdict(set)
+    el_by_cpe: defaultdict[str, list] = defaultdict(list)
```

Comment:
Maybe el_suffix_by_cpe is a better name:
```suggestion
    el_suffix_by_cpe: defaultdict[str, list] = defaultdict(list)
```

Reply:
updated